### ISSUE TITLE
CallableFunction inference does not match runtime behavior

### ISSUE THREAD

**Issue by karol-majewski** [Microsoft/TypeScript#30294]
**TypeScript Version:**  3.4.0-dev.20190309

**Search Terms:** CallableFunction, call, bind, inference

**Code**

```ts
Promise.resolve.call(1)           // Compiles without error, throws in runtime
Promise.resolve.call(Promise, 1); // Compile-time error, works fine in runtime
```

**Expected behavior:**

- `Promise.resolve.call(1)` should trigger a compile-time error
- `Promise.resolve.call(Promise, 1)` should compile

**Actual behavior:**

The situation is reversed. The one that works in runtime does not compile. The one that _does_ compile throws a runtime error.

- `Promise.resolve.call(1)` compiles without error, throws `Uncaught TypeError: PromiseResolve called on non-object` in runtime
- `Promise.resolve.call(Promise, 1)` gives a compile-time error (`Expected 1 arguments, but got 2.`), works fine in runtime (returns a `Promise<number>`)

I'm using `"lib": ["dom", "esnext"], "target: "es2015"`. Tested on Chrome 72.0.3626.121, V8 7.2.502.28.

**Playground Link:**  [The compile-time error is not there, but the runtime exception can be observed](https://www.typescriptlang.org/play/#src=Promise.resolve.call(1)%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20Compiles%20without%20error%2C%20throws%20in%20runtime%0D%0APromise.resolve.call(Promise%2C%201)%3B%20%2F%2F%20Compile-time%20error%2C%20works%20fine%20in%20runtime)

**Comment by RyanCavanaugh** [comment]
@rbuckton bug or no?

**Comment by rbuckton** [comment]
@RyanCavanaugh: This is likely due to overload resolution related to `--strictBindCallApply`, as we see the same error in this case:

```ts
declare class C{
  static x(y: number): number;
  static x(): number;
}
C.x.call(C, 1); // error, but should be ok
C.x.call(C); // ok
C.x.call(1); // ok (not an error because the `this` type of `x` is not constrained)
```

**Comment by rbuckton** [comment]
Note that if you switch the order of the overloads, you get the exact opposite results (`C.x.call(c, 1)` works but `C.x.call(C)` and `C.x.call(1)` fail).

**Comment by rbuckton** [comment]
@RyanCavanaugh: Not a bug per se, as we have [documented this is the expected behavior](www.typescriptlang.org/docs/handbook/release-notes/typescript-3-2.html#caveats), however it does result in a poor user experience.
